### PR TITLE
Adds base for test cluster authorization

### PR DESCRIPTION
**Pull request by chentex**
Refactors list clusters test to re-use code
Adds option WriteReport for all tests

Signed-off-by: Vicente Zepeda Mas <vzepedam@redhat.com>

### CODE REVIEW

File: pkg/tests/tests.go [hunk 1]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
 	}
 	return buff.Bytes()
 }
+
+func clusterAuthorizationsBody() []byte {
```
Comment (chentex) [inline pkg/tests/tests.go:121]: @nimrodshn could you help me with this part. I have no idea what parameters values we need to use to build the body, or if we can build it here, maybe it has to be built individually for each request we want to try.

File: pkg/tests/tests.go [hunk 2]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+	buff := &bytes.Buffer{}
+	reservedResource := acctmgmtv1.NewReservedResource()
+
+	clusterAuthReq, err := acctmgmtv1.NewClusterAuthorizationRequest().
+		AccountUsername(helpers.AccountUsername).
+		AvailabilityZone("???").
```
Comment (markturansky) [inline pkg/tests/tests.go:127]: "single" or "multi"

File: pkg/tests/tests.go [hunk 3]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		BYOC(true).
+		CloudAccountID("???").
```
Comment (markturansky) [inline pkg/tests/tests.go:129]: can omit

File: pkg/tests/tests.go [hunk 4]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		CloudProviderID("???").
```
Comment (markturansky) [inline pkg/tests/tests.go:130]: "aws" or "gcp"

File: pkg/tests/tests.go [hunk 5]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		ClusterID("???").
```
Comment (markturansky) [inline pkg/tests/tests.go:131]: arbitrary id

File: pkg/tests/tests.go [hunk 6]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		Disconnected(true).
+		DisplayName("???").
```
Comment (markturansky) [inline pkg/tests/tests.go:133]: can omit

File: pkg/tests/tests.go [hunk 7]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		ExternalClusterID("???").
```
Comment (markturansky) [inline pkg/tests/tests.go:134]: arbitrary ID

File: pkg/tests/tests.go [hunk 8]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		Managed(true).
```
Comment (markturansky) [inline pkg/tests/tests.go:135]: not used, to be honest

File: pkg/tests/tests.go [hunk 9]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		ProductCategory("assistedInstall").
```
Comment (markturansky) [inline pkg/tests/tests.go:136]: "AssistedInstall" but should be case-insentive

File: pkg/tests/tests.go [hunk 10]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		ProductID("osd").
```
Comment (markturansky) [inline pkg/tests/tests.go:137]: AssistedInstall is a category only of OCP plans, not OSD.

File: pkg/tests/tests.go [hunk 11]
```diff
@@ -108,3 +117,35 @@ func resourceReviewbody() []byte {
...
+		Reserve(true).
+		Resources(reservedResource).
```
Comment (markturansky) [inline pkg/tests/tests.go:139]: OCP plans do not require resources to be sent.  OSD plans require resources.

See our integration test of OCP AssistedInstaller:  https://gitlab.cee.redhat.com/service/uhc-account-manager/-/blob/master/test/integration/cluster_authorization_test.go#L3206

File: pkg/tests/tests.go (hunk 2 above)
Comment (markturansky) [reply pkg/tests/tests.go:127]: "single", based on the MW00530 sku granted to your org

Comment (markturansky) [reply pkg/tests/tests.go:127]: but if OCP, neither value really matters as all OCP maps to 0 cost.  See row 25:

https://docs.google.com/spreadsheets/d/1HGvQnahZCxb_zYH2kSnnTFsxy9MM49vywd-P0X_ISLA/edit#gid=968570191

File: README.md [hunk 12]
```diff
@@ -65,6 +67,7 @@ Each test can contain this options:
 - freq: Number of requests to execute in a unit of time.
 - per: Unit of the request frequency. ("ns", "us" (or "µs"), "ms", "s", "m", "h")
 - duration: Override duration for the test. (A positive integer accompanied of a valid unit)
+- writeReport: Override write report option for the test. ("true" or "fasle")
```
Comment (nimrodshn) [inline README.md:70]: /s/fasle/false

File: pkg/tests/tests.go (hunk 1 above)
Comment (nimrodshn) [reply pkg/tests/tests.go:121]: @markturansky for save! He definitely knows better for the SDB endpoints.

**Comment by nimrodshn** [comment]
@markturansky AFAIK `ClusterAuthorization` creates a subscription record in the AMS DB, what other side effects can be triggered by this test? is there a mechanism to make this a "fake" call? b/c we would not need these subscription post this test and wouldn't want to pollute even the integration envs DB if we hit this endpoint very hard. WDYT?

File: pkg/tests/handlers/accounts.go [hunk 13]
```diff
@@ -125,3 +133,78 @@ func TestQuotaCost(options *helpers.TestOptions) error {
...
+		ClusterID(clusterID).
+		ProductID("osd").
+		CloudProviderID("aws").
+		AccountUsername(helpers.AccountUsername).
+		Managed(true).
+		Reserve(true).
+		BYOC(true).
+		AvailabilityZone("single").
+		Resources(reservedResource).
+		Build()
```
Comment (nimrodshn) [inline pkg/tests/handlers/accounts.go:199]: @chentex Sorry for nitpicking do you mind making all of these strings into `consts` somewhere?

**Comment by nimrodshn** [comment]
@chentex Looks very good, some nitpicking but all in all :+1:

File: pkg/helpers/constants.go [hunk 14]
```diff
@@ -1,5 +1,19 @@
 package helpers
 
 const (
-	AccountUsername = "rhn-support-tiwillia"
+	AccountUsername  = "rhn-support-tiwillia"
+	DefaultAWSRegion = "us-east-1"
+
+	// Cluster Auhtorization
+	ClusterAuthResourceName     = "m5.xlarge"
```
Comment (nimrodshn) [inline pkg/helpers/constants.go:8]: I would name this 'ResourceNameM5XLarge`,

File: pkg/helpers/constants.go [hunk 15]
```diff
@@ -1,5 +1,19 @@
...
+	ClusterAuthResourceType     = "compute.node.aws"
+	ClusterAuthCount            = 4
+	ClusterAuthBillingModel     = "standard"
```
Comment (nimrodshn) [inline pkg/helpers/constants.go:11]: What about `StandardBillingModel` ?

File: pkg/helpers/constants.go [hunk 16]
```diff
@@ -1,5 +1,19 @@
 package helpers
 
 const (
-	AccountUsername = "rhn-support-tiwillia"
+	AccountUsername  = "rhn-support-tiwillia"
+	DefaultAWSRegion = "us-east-1"
+
+	// Cluster Auhtorization
+	ClusterAuthResourceName     = "m5.xlarge"
+	ClusterAuthResourceType     = "compute.node.aws"
```
Comment (nimrodshn) [inline pkg/helpers/constants.go:9]: Perhaps `AWSComputeNodeResourceType`.

File: pkg/helpers/constants.go [hunk 17]
```diff
@@ -1,5 +1,19 @@
...
+	ClusterAuthCount            = 4
```
Comment (nimrodshn) [inline pkg/helpers/constants.go:10]: Perhaps `DefaultClusterCount`?

File: pkg/helpers/constants.go [hunk 18]
```diff
@@ -1,5 +1,19 @@
...
+	ClusterAuthBillingModel     = "standard"
+	ClusterAuthProductID        = "osd"
+	ClusterAuthCloudProviderID  = "aws"
```
Comment (nimrodshn) [inline pkg/helpers/constants.go:13]: Perhpas `AWSCloudProvider`?

File: pkg/helpers/constants.go [hunk 19]
```diff
@@ -1,5 +1,19 @@
...
+	ClusterAuthAccountUsername  = "rh-perfscale"
+	ClusterAuthManaged          = true
+	ClusterAuthReserve          = true
+	ClusterAuthBYOC             = true
+	ClusterAuthAvailabilityZone = "single"
```
Comment (nimrodshn) [inline pkg/helpers/constants.go:18]: Perhaps `SingleAvailabilityZone`.

**Review by nimrodshn — COMMENTED** [review]
Just a little more small nits .. @chentex 

I would like the names to relate to the contents.. We might have more tests with `gcp` cloudprovider and `multi` zone availability, etc..

So using `ClusterAuthAvailabilityZone` for availability doesn't reflect it meaning `single`..

**Comment by chentex** [comment]
Thank you @nimrodshn makes sense to have more descriptive names :+1:

**Review by nimrodshn — APPROVED** [review]
Very nice, LGTM :100:

**Review by kurtismullins — APPROVED** [review]
LGTM

**Pull request merged**